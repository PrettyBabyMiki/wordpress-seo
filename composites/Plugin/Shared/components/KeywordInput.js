// External dependencies.
import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import uniqueId from "lodash/uniqueId";
import { __ } from "@wordpress/i18n";

// Internal dependencies.
import colors from "../../../../style-guide/colors.json";

const errorColor = colors.$color_red;

const KeywordInputContainer = styled.div`
	display: flex;
	flex-direction: column;
	margin: 1em 0;
`;

const KeywordFieldLabel = styled.label`
	font-size: 1em;
	font-weight: bold;
	margin-bottom: 0.5em;
`;

const KeywordField = styled.input`
	border: 3px solid initial;
	padding: 0.75em;
	font-size: 1em;

	&.hasError {
		border-color: ${ errorColor };
		outline-color: ${ errorColor };
	}
`;

const ErrorText = styled.p`
	color: ${ errorColor };
	margin: 0.5em 0 0 0;
	min-height: 1.8em;
`;

class KeywordInput extends React.Component {
	/**
	 * Constructs a KeywordInput component.
	 *
	 * @param {Object}   props           The props for this input field component.
	 * @param {string}   props.id        The id of the KeywordInput.
	 * @param {string}   props.label     The label of the KeywordInput.
	 * @param {boolean}  props.showLabel Toggle between an actual label or an aria-label on the input.
	 * @param {string}   props.keyword   The initial keyword passed to the state.
	 * @param {function} props.onChange  The function that is triggered when the keyword input field is changed.
	 *
	 * @returns {void}
	 */
	constructor( props ) {
		super( props );

		this.handleChange = this.handleChange.bind( this );
		this.displayErrorMessage = this.displayErrorMessage.bind( this );
	}

	/**
	 * Checks the keyword input for comma-separated words.
	 *
	 * @param {string} keywordText The text of the input.
	 *
	 * @returns {boolean} Returns true if a comma was found.
	 */
	checkKeywordInput( keywordText ) {
		return keywordText.includes( "," );
	}

	/**
	 * Displays the error message
	 *
	 * @param {boolean} showErrorMessage Whether or not the error message has to be shown.
	 *
	 * @returns {ReactElement} ErrorText The error message element.
	 */
	displayErrorMessage( showErrorMessage ) {
		if ( showErrorMessage && this.props.keyword !== "" ) {
			return (
				<ErrorText role="alert">
					{ __( "Are you trying to use multiple keywords? You should add them separately below.", "yoast-components" ) }
				</ErrorText>
			);
		}
	}

	/**
	 * Handles changes in the KeywordInput.
	 *
	 * @param {SyntheticEvent} event The onChange event.
	 *
	 * @returns {void} Sets the state if a change has been made.
	 */
	handleChange( event ) {
		this.props.onChange( event.target.value );
	}

	/**
	 * Renders an input field, a label, and if the condition is met, an error message.
	 *
	 * @returns {ReactElement} The KeywordField react component including its label and eventual error message.
	 */
	render() {
		const { id, label, showLabel, keyword } = this.props;
		const showErrorMessage = this.checkKeywordInput( keyword );

		return(
			<KeywordInputContainer>
				{ showLabel && <KeywordFieldLabel htmlFor={ id }>
					{ label }
				</KeywordFieldLabel> }
				<KeywordField
					aria-label={ showLabel ? null : label }
					type="text"
					id={ id }
					className={ showErrorMessage ? "hasError" : null }
					onChange={ this.handleChange }
					value={ keyword }
				/>
				{ this.displayErrorMessage( showErrorMessage ) }
			</KeywordInputContainer>
		);
	}
}

export const keywordInputPropType = {
	id: PropTypes.string,
	label: PropTypes.string.isRequired,
	showLabel: PropTypes.bool,
	keyword: PropTypes.string,
	onChange: PropTypes.func.isRequired,
};

KeywordInput.propTypes = keywordInputPropType;

KeywordInput.defaultProps = {
	id: uniqueId( "yoast-keyword-input-" ),
	showLabel: true,
	label: __( "Focus keyword:", "yoast-components" ),
	keyword: "",
};

export default KeywordInput;

