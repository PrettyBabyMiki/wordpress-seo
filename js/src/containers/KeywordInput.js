import { connect } from "react-redux";
import KeywordInput from "yoast-components/composites/Plugin/Shared/components/KeywordInput";
import { setFocusKeyword } from "../redux/actions/focusKeyWord";

/**
 * Maps redux state to KeywordInput props.
 *
 * @param {Object} state The redux state.
 *
 * @returns {Object} Props that should be passed to SeoAnalysis.
 */
function mapStateToProps( state ) {
	return {
		keyword: state.focusKeyword,
	};
}

/**
 * Maps the redux dispatch to KeywordInput props.
 *
 * @param {Function} dispatch The dispatch function that will dispatch a redux action.
 *
 * @returns {Object} Props for the `KeywordInput` component.
 */
function mapDispatchToProps( dispatch ) {
	return {
		onChange: ( value ) => {
			dispatch( setFocusKeyword( value ) );
		},
	};
}

export default connect( mapStateToProps, mapDispatchToProps )( KeywordInput );
