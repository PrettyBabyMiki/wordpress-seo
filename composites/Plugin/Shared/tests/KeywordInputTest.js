import React from "react";
import renderer from "react-test-renderer";
import EnzymeAdapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme/build/index";

import KeywordInput from "../components/KeywordInput";

Enzyme.configure( { adapter: new EnzymeAdapter() } );

describe( KeywordInput, () => {
	it( "matches the snapshot by default", () => {
		const component = renderer.create(
			<KeywordInput id="test-id" onChange={ () => {
			} } label="test label"/>
		);

		let tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "does not display the error message for a single keyword", () => {
		const wrapper = Enzyme.mount(
			<KeywordInput id="test-id" label="test label" />
		);
		wrapper.find( "input" ).simulate( "change", {
			target: {
				value: "Keyword1",
			},
		} );
		expect( wrapper.state().showErrorMessage ).toEqual( false );
	} );

	it( "does not display the error message for two words separated by whitespace", () => {
		const wrapper = Enzyme.mount(
			<KeywordInput id="test-id" label="test label" />
		);
		wrapper.find( "input" ).simulate( "change", {
			target: {
				value: "Keyword1 Keyword2",
			},
		} );
		expect( wrapper.state().showErrorMessage ).toEqual( false );
	} );

	it( "displays the error message for comma-separated words", () => {
		const wrapper = Enzyme.mount(
			<KeywordInput id="test-id" label="test label" />
		);
		wrapper.find( "input" ).simulate( "change", {
			target: {
				value: "Keyword1, Keyword2",
			},
		} );
		expect( wrapper.state().showErrorMessage ).toEqual( true );
		expect( wrapper.find( "div" ).text() ).toBeTruthy();
	} );
} );
