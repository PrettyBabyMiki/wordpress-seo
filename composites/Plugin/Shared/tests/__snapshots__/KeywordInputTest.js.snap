// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeywordInput matches the snapshot by default 1`] = `
.c5 {
  border: 1px solid #ddd;
  padding: 0.75em;
  font-size: 1em;
}

.c7::-moz-focus-inner {
  border-width: 0;
}

.c7:focus {
  outline: none;
  border-color: #0066cd;
  color: #000;
  background-color: #fff;
  box-shadow: 0 0 3px rgba( 8,74,103,0.8 );
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 1em 0;
}

.c1 {
  font-size: 1em;
  font-weight: bold;
  margin-bottom: 0.5em;
}

.c4 {
  -webkit-flex: 1 !important;
  -ms-flex: 1 !important;
  flex: 1 !important;
  box-sizing: border-box;
  max-width: 100%;
  margin: 0;
}

.c4.has-error {
  border-color: #dc3232 !important;
  box-shadow: 0 0 2px #dc3232 !important;
}

.c8 {
  border: 1px solid transparent;
  box-shadow: none;
  background: none;
  -webkit-flex: 0 0 32px;
  -ms-flex: 0 0 32px;
  flex: 0 0 32px;
  height: 32px;
  max-width: 32px;
  padding: 0;
  cursor: pointer;
}

.c9 {
  margin-top: 4px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2.has-remove-keyword-button .c3 {
  padding-right: 40px;
}

.c2.has-remove-keyword-button .c6 {
  margin-left: -32px;
}

.c10 {
  width: 18px;
  height: 18px;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="test-id"
  >
    Focus keyword:
  </label>
  <div
    className="has-remove-keyword-button c2"
  >
    <input
      aria-label={null}
      className="c3 c4 c5"
      id="test-id"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="c6 c7 c8"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="yoast-svg-icon yoast-svg-icon-times-circle c9 c10"
        fill="#767676"
        focusable="false"
        role="img"
        size="18px"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
        />
      </svg>
    </button>
  </div>
</div>
`;
